kubelet: Panic on portforward streams

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!

If the matter is security related, please disclose it privately via https://kubernetes.io/security/
-->


#### What happened:

Since upgrading to Kubernetes v1.19.11 (from v1.19.10) we see occasional panics in kubelet related to portforward streams:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x17e78f8]
goroutine 1168860 [running]:
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/httpstream/spdy.(*connection).RemoveStreams(0xc002b30080, 0xc0032f7160, 0x2, 0x2)
 /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/httpstream/spdy/connection.go:101 +0x78
k8s.io/kubernetes/pkg/kubelet/cri/streaming/portforward.(*httpStreamHandler).removeStreamPair(0xc00349c310, 0x6c657a8, 0x1)
 /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/cri/streaming/portforward/httpstream.go:168 +0x16d
k8s.io/kubernetes/pkg/kubelet/cri/streaming/portforward.(*httpStreamHandler).monitorStreamPair(0xc00349c310, 0xc000f6ef50, 0xc00369fd40)
 /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/cri/streaming/portforward/httpstream.go:148 +0x1fe
created by k8s.io/kubernetes/pkg/kubelet/cri/streaming/portforward.(*httpStreamHandler).run
 /workspace/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/cri/streaming/portforward/httpstream.go:227 +0x58a
 ```

This issue seem to be introduced with https://github.com/kubernetes/kubernetes/pull/99839


#### What you expected to happen:

`kubelet` should not panic and crash.

#### How to reproduce it (as minimally and precisely as possible):

We have not been able to successfully force this to happen, but it has happened 41 times in the last 2 weeks across 7 clusters (~100 nodes).

#### Anything else we need to know?:

#### Environment:
- Kubernetes version (use `kubectl version`):

```
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.11", GitCommit:"c6a2f08fc4378c5381dd948d9ad9d1080e3e6b33", GitTreeState:"clean", BuildDate:"2021-05-12T12:19:22Z", GoVersion:"go1.15.12", Compiler:"gc", Platform:"linux/amd64"}
```

- Cloud provider or hardware configuration: Self managed clusters on AWS.
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

/sig node

/cc @saschagrunert

--- a/staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/connection.go
+++ b/staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/connection.go
@@ -123,7 +123,10 @@ func (c *connection) Close() error {
 func (c *connection) RemoveStreams(streams ...httpstream.Stream) {
 	c.streamLock.Lock()
 	for _, stream := range streams {
-		delete(c.streams, stream.Identifier())
+		// It may be possible that the provided stream is nil if timed out.
+		if stream != nil {
+			delete(c.streams, stream.Identifier())
+		}
 	}
 	c.streamLock.Unlock()
 }

--- a/staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/connection_test.go
+++ b/staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/connection_test.go
@@ -323,6 +323,9 @@ func TestConnectionRemoveStreams(t *testing.T) {
 	// remove all existing
 	c.RemoveStreams(stream0, stream1)
 
+	// remove nil stream should not crash
+	c.RemoveStreams(nil)
+
 	if len(c.streams) != 0 {
 		t.Fatalf("should not have any streams, has %d", len(c.streams))
 	}

